src/persist-client/src/internal/trace.rs

```diff
@@ -103,10 +112,225 @@ impl<T: Timestamp + Lattice> Default for Trace<T> {
     fn default() -> Self {
         Self {
             spine: Spine::new(),
+            roundtrip_structure: false,
         }
     }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct ThinSpineBatch<T> {
+    pub(crate) level: usize,
+    pub(crate) desc: Description<T>,
+    pub(crate) parts: Vec<SpineId>,
+}
+
+#[derive(Clone, Debug)]
+pub struct FlatTrace<T> {
+    pub(crate) since: Antichain<T>,
+    pub(crate) legacy_batches: BTreeMap<Arc<HollowBatch<T>>, ()>,
+    pub(crate) hollow_batches: BTreeMap<SpineId, Arc<HollowBatch<T>>>,
+    pub(crate) spine_batches: BTreeMap<SpineId, ThinSpineBatch<T>>,
+    pub(crate) fueling_merges: BTreeMap<SpineId, FuelingMerge<T>>,
+}
+impl<T: Timestamp + Lattice> Trace<T> {
+    pub(crate) fn flatten(&self) -> FlatTrace<T> {
+        let since = self.spine.since.clone();
+        let mut legacy_batches = BTreeMap::new();
+        // NB: for backwards-compatability reasons, we don't write these out yet!
```

Comment:
nit: s/don't write these/don't write hollow_batches/ (I misunderstood this comment the first time I read it to also be talking about spine_batches and fueling_merged)